### ISSUE TITLE
[REVIEW] Unnecessary allocation of memory?

### ISSUE THREAD

**Issue by santisoler** [brmather/pycurious#17]
Part of https://github.com/openjournals/joss-reviews/issues/1544.

Is there a special reason why `data` is copied on the following line?
https://github.com/brmather/pycurious/blob/759fc520c22503f0b40b6a63faa4e632dd9f1e88/pycurious/grid.py#L180

Remember that the `copy()` method allocates memory, what would increase the memory consumption if `data` is too big and require more computation time.

**Comment by brmather** [comment]
Yes this was a conscious decision to ensure that `data` is not overwritten in place. The alternative is to replace

```python
data.flat[:] -= np.dot(A, c)
```
with
```python
data_new = data.flat[:] - np.dot(A,c)
return data_new.reshape(data.shape)
```
but that's no more efficient.

However, revisiting this method was good because I replaced `np.meshgrid` with the more efficient `np.mgrid` function.

**Comment by santisoler** [comment]
There's no special need for defining `data_new`: it is just going to be returned on the next line. So I think this should work:
```python
def remove_trend_linear(self, data):
    """
    Remove the best-fitting linear trend from the data
    This may come in handy if the magnetic data has not been
    reduced to the pole.
    Args:
        data : 2D numpy array
    Returns:
     data : 2D numpy array
    """
    nr, nc = data.shape
    yq, xq = np.mgrid[0:nc,0:nr]
    A = np.c_[xq.ravel(), yq.ravel(), np.ones(xq.size)]
    c, resid, rank, sigma = np.linalg.lstsq(A, data.ravel(), rcond=None)
    return data - np.dot(A, c).reshape(data.shape)
```

**Comment by brmather** [comment]
Thanks @santisoler I have re-implemented this function accordingly in f3adb03daab9debdc739130c5caba5bc213c4789.
